flocker/node/agents/cinder.py

```diff
@@ -67,30 +71,42 @@ def _run_with_logging(self, *args, **kwargs):
 
         # See https://clusterhq.atlassian.net/browse/FLOC-2054
         # for ensuring all method arguments are serializable.
-        with OPENSTACK_ACTION(operation=[method_name, args, kwargs]):
-            try:
-                return method(*args, **kwargs)
-            except NovaClientException as e:
-                NOVA_CLIENT_EXCEPTION(
-                    code=e.code,
-                    message=e.message,
-                    details=e.details,
-                    request_id=e.request_id,
-                    url=e.url,
-                    method=e.method,
-                ).write()
-                raise
-            except KeystoneHttpError as e:
-                KEYSTONE_HTTP_ERROR(
-                    code=e.http_status,
-                    message=e.message,
-                    details=e.details,
-                    request_id=e.request_id,
-                    url=e.url,
-                    method=e.method,
-                    response=e.response.text,
-                ).write()
-                raise
+        retry = True
+        attempt = 0
+        while (retry and attempt < MAX_ATTEMPTS):
+            with OPENSTACK_ACTION(operation=[attempt, method_name, args, kwargs]):
+                try:
+                    return method(*args, **kwargs)
+                except KeystoneOverLimit as e:
+                    retry_after = int(e.retry_after)
+                    time.sleep(retry_after)
+                    retry = True
+                    attempt += 1
+                    if attempt == MAX_ATTEMPTS:
+                        raise
```

Comment:
Thanks! Unit test is a good idea! This is how i reproduced / tested fix:

```
    def test_overlimit(self):
        """
        ``attach_volume`` can attach multiple block devices to a single host.
        """

        self.assertRaises(KeystoneOverLimit, self._generate_overlimit)

    def _generate_overlimit(self):
        i = 0
        while i < 50:
            self.api.create_volume(
                dataset_id=uuid4(),
                size=self.minimum_allocatable_size
            )
            i += 1
```

Did not include it in the review because it doesn't belong in functional test suite.

Unit test is a better idea, will move to that.

i also wanted review comments on using ``retry_after`` primarily versus max(retry_after, exponential back off generated value). The few reproductions i saw, ``retry_after`` has always been ``0``. Hence, i do not completely trust relying solely on that. What do you think, @exarkun ?